### PR TITLE
Feat/new directions

### LINKED ISSUE #115: Direction issues when using Pointy Orientation
**Summary**
The angle values listed in the `Direction` documentation for pointy orientation are off by one (`South` should be pointy angle 0), as the documented angle values for pointy orientation do not match the coordinate change described for each Direction.

On top of this, there's a larger issue of lacking Pointy orientation `Direction` as directions such as `NorthWest` have different coordinate changes for each orientation.

**Details**

The `Direction` enum uses flat orientation for its naming, but the documentation for each direction states what equivalent direction it is for pointy orientation.  For example `Direction::TopRight` notes that in pointy orientation, it's equivalent of 0 radians, which would be East (assuming mathematical east-counterclockwise orientation).

The problem is, the coordinate change of `TopRight` in flat orientation is `(1, -1)`, but the change for `East` in pointy orientation is `(1, 0)` (which is `Bottom` in flat orientation).

This becomes a problem when you start using `Hex::all_neighbors()` using pointy orientation, which actually starts at `NorthEast` instead of `East`.

This matters when storing data by `Direction` index (`Direction as usize`), as `Direction::NorthEast` acting as `PointyDirection::East` has index 0, but the actual transform `(1, -1)` leads to the cell to the north-east in pointy orientation.

**To Reproduce**
According to the documents `Direction::TopRight` is supposed to be equivalent to `East` when using pointy orientation, so this should pass:
```rust
pub type East = Direction::TopRight;
assert_eq!(Hex::new(0, 0) + East, Hex::new(0, 1));
```

**Expected behavior**
Honestly, I don't know.  Updating the documentation to change the angles to match the
direction coordinate changes in `NEIGHBORS_COORDS` would be a start.  Effectively `Direction::South` should be pointy angle 0.

However, this doesn't really resolve the overarching problem of having to mentally convert  the pointy direction `East` into flat direction `South`.  It gets even worse when you realize `NorthWest` in flat `(-1, 1)` is different from `NorthWest` in pointy `(0, -1)`.

Perhaps specialization of `Direction` into `PointyDirection` and `FlatDirection` is the approach?

**Screenshots Or Code snippet**
Pointy orientation coordinate change:
Screen Shot 2023-09-10 at 12 51 12 PM

Flat orientation coordinate change:
Screen Shot 2023-09-10 at 12 51 24 PM

### PR DESCRIPTION
**Pull request by ManevilleF**
> Closes #115

## Problem

The direction enums, `Direction` and `DiagonalDirection` were confusing users, because their variants had orientation based names.

For example `Direction::Top` made sense in flat orientation, but in pointy orientation with the 30 degrees shift, it lost its meaning.
Multiple users thought directions were related to `HexOrientation` when they are not, due to this unfortunate naming.

## Solution

First the `Direction` enum was renamed to `EdgeDirection` , and `DiagonalDirection` to `VertexDirection`. This naming is more consistent on what those directions actually represent.

- `EdgeDirection` represents the vector to reach an _edge-adjacent_ neigbor of a given coordinate

```txt
           ___
          /   \
      +--+  1  +--+
     /  2 \___/ 0  \
     \    /   \    /
      +--+     +--+
     /  3 \___/ 5  \
     \    /   \    /
      +--+  4  +--+  
          \___/
 ```

- `VertexDirection` represents the vector to reach a diagonal neibhor following the vertex/corner of the hexagon

```txt
          \___/
     \ 2  /   \ 1  /
      +--+     +--+
   __/    \___/    \__
     \    /   \    /
   3  +--+     +--+  0
   __/    \___/    \__
     \    /   \    /
      +--+     +--+ 
     / 4  \___/  5 \
 ```
 
 Then, instead of an enum, which forced me to name the directions, they are now structs storing a `u8` index between 0 and 5, 6 possibilities. The enum way was also `repr(u8)` mimicking this index way
 
 To keep the ability to use directions directly with orientation based names I added a **bunch** of const values for both _flat_ and _pointy_ orientation
 
 ## Notes
 
 If this PR is a huge breaking change, there should not be any lost feature, I kept all methods available before to both direction enums, and all the tests were adapted
 
 The idea for the renaming comes from #154 , where I realized that directions are edge related, and diagonals vertex related

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Oh I like that rename: I think it's much clearer!

**Review by alice-i-cecile — APPROVED** [review]

File: src/hex/mod.rs [hunk 1]
```diff
@@ -1086,3 +1097,13 @@ impl Hex {
         self.to_local(range)
     }
 }
+
+impl Debug for Hex {
```
Comment (ManevilleF) [inline src/hex/mod.rs:1101]:
```suggestion
#[cfg(not(target_arch = "spirv"))]
impl Debug for Hex {
```

File: src/direction/edge_direction.rs [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+///
+/// Example:
+/// ```rust
+/// # use hexx::*;
+/// let direction = EdgeDirection::FLAT_TOP;
+/// assert_eq!(-direction, EdgeDirection::FLAT_BOTTOM);
+/// assert_eq!(direction >> 1, EdgeDirection::FLAT_TOP_RIGHT);
+/// assert_eq!(direction << 1, EdgeDirection::FLAT_TOP_LEFT);
+/// ```
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Default)]
```
Comment (ManevilleF) [inline src/direction/edge_direction.rs:49]:
```suggestion
#[derive(Clone, Copy, PartialEq, Eq, Default)]
#[cfg_attr(not(target_arch = "spirv"), derive(Hash))]
```

File: src/direction/vertex_direction.rs [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+///
+/// Example:
+/// ```rust
+/// # use hexx::*;
+/// let direction = VertexDirection::FLAT_RIGHT;
+/// assert_eq!(-direction, VertexDirection::FLAT_LEFT);
+/// assert_eq!(direction >> 1, VertexDirection::FLAT_BOTTOM_RIGHT);
+/// assert_eq!(direction << 1, VertexDirection::FLAT_TOP_RIGHT);
+/// ```
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Default)]
```
Comment (ManevilleF) [inline src/direction/vertex_direction.rs:46]:
```suggestion
#[derive(Clone, Copy, PartialEq, Eq, Default)]
#[cfg_attr(not(target_arch = "spirv"), derive(Hash))]
```

File: src/hex/mod.rs (hunk 1 above)
Comment (alice-i-cecile) [reply src/hex/mod.rs:1101]: Interesting; why is this needed?

Comment (ManevilleF) [reply src/hex/mod.rs:1101]: Deubg and Hash are not supported on spirv IIRC, also on the GPU we don't care much about `Debug`

**Pull request merged**